### DISCUSSION TITLE
Add an auth log to enable the use of tools like fail2ban

### DISCUSSION THREAD

**Discussion by netw0rk-noob** [firefly-iii/firefly-iii#9014] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/explanation/more-information/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been requested before.

### Description

I would like Firefly to add the option to log (failed) authentication attempts. As it contains (possibly sensitive) financial data I deem it a security risk that its currently possible to infinitely try guessing (bruteforcing) logins on the webinterface.

### Solution

Adding an option to enable authentication logs (containing the source ip addresses of failed login attempts) to a specific file (or systemd) would create the possibillity to automatically ban every ip address which failed to authenticate `n` times during a defined period of time using tools like fail2ban. That would drastically reduce the risk of unauthenticated access to Firefly.

### What are alternatives?

I already [enabled debug logging](https://docs.firefly-iii.org/how-to/general/debug/) but it didnt contain login attempts either.

A theoretical alternative would be to enable HTTP basic auth in the webserver (which would add a cumbersome second login in front of the Firefly login) and use fail2ban on its logs.

**Comment by JC5** [comment, marked as answer]
Check out the audit logs. From `.env.example`

```
# Audit log level.
# The audit log is used to log notable Firefly III events on a separate channel.
# These log entries may contain sensitive financial information.
# The audit log is disabled by default.
#
# To enable it, set AUDIT_LOG_LEVEL to "info"
# To disable it, set AUDIT_LOG_LEVEL to "emergency"
AUDIT_LOG_LEVEL=emergency
```

**Comment by netw0rk-noob** [comment]
Thank you for this very fast reply. That does indeed enable the functionallity I'm after, as it produces logs like
`production.WARNING: AUDIT: Login failed. Attempt for user "user@domain.tld" failed. (XXX.XXX.XXX.XXX -> POST:https://firefly-iii.tld/login)`
It even works if I set `AUDIT_LOG_LEVEL=warning`.

I have no idea why I didnt think of trying this. Its just beneath `APP_LOG_LEVEL` which I tried to change.
Thank you very much for clearing this up for me!

**Comment by netw0rk-noob** [comment]
I can confirm: This log works perfectly with fail2ban. Sorry again to have bothered you with my blindness, @JC5. And thank you you very much for developing this very helpful tool!

Here is my (probably far from perfect) `/etc/fail2ban/filter.d/firefly-iii-audit.conf`, just in case anyone wants to use it or build a better one from it:
```
[Definition]
failregex = .*production.WARNING: AUDIT: Login failed. Attempt for user \"[^\"]+\" failed. \(<HOST> -> [^)]+\)
ignoreregex =
```

**Reply by JC5** [reply]
Not a problem, thanks for the compliments!
